owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/oauth/TokenRequestRemoteOperation.kt

```diff
@@ -0,0 +1,87 @@
+ *   all copies or substantial portions of the Software.
+ *
+ *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ *   BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ *   ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ *   CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *   THE SOFTWARE.
+ *
+ */
+package com.owncloud.android.lib.resources.oauth
+
+import com.owncloud.android.lib.common.OwnCloudClient
+import com.owncloud.android.lib.common.http.HttpConstants.AUTHORIZATION_HEADER
+import com.owncloud.android.lib.common.http.HttpConstants.HTTP_OK
+import com.owncloud.android.lib.common.http.methods.nonwebdav.PostMethod
+import com.owncloud.android.lib.common.operations.RemoteOperation
+import com.owncloud.android.lib.common.operations.RemoteOperationResult
+import com.owncloud.android.lib.resources.oauth.params.TokenRequestParams
+import com.owncloud.android.lib.resources.oauth.responses.TokenResponse
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.Moshi
+import timber.log.Timber
+import java.net.URL
+
+/**
+ * Perform token request
+ *
+ * @author Abel García de Prada
+ */
+class TokenRequestRemoteOperation(
+    private val tokenRequestParams: TokenRequestParams
+) : RemoteOperation<TokenResponse>() {
+
+    override fun run(client: OwnCloudClient): RemoteOperationResult<TokenResponse> {
+        try {
+            val requestBody = tokenRequestParams.toRequestBody()
+
+            val postMethod = PostMethod(URL(tokenRequestParams.tokenEndpoint), requestBody)
+
+            postMethod.addRequestHeader(AUTHORIZATION_HEADER, tokenRequestParams.clientAuth)
+
+            val status = client.executeHttpMethod(postMethod)
+
+            val responseBody = postMethod.getResponseBodyAsString()
+
+            if (status == HTTP_OK && responseBody != null) {
+                Timber.d("Successful response $responseBody")
+
+                // Parse the response
+                val moshi: Moshi = Moshi.Builder().build()
+                val jsonAdapter: JsonAdapter<TokenResponse> = moshi.adapter(TokenResponse::class.java)
+                val tokenResponse: TokenResponse? = jsonAdapter.fromJson(responseBody)
+                Timber.d("Get tokens completed and parsed to $tokenResponse")
+
+                return RemoteOperationResult<TokenResponse>(RemoteOperationResult.ResultCode.OK).apply {
+                    data = tokenResponse
+                }
+
+            } else {
+                Timber.e("Failed response while getting tokens from the server status code: $status; response message: $responseBody")
+                return RemoteOperationResult<TokenResponse>(postMethod)
+            }
```

Comment:
```suggestion
with(tokenRequestParams) {
    PostMethod(
        url = URL(tokenEndpoint),
        postRequestBody = toRequestBody()
    ).apply {
        addRequestHeader(AUTHORIZATION_HEADER, clientAuth)
    }.run {
        val status = client.executeHttpMethod(this)
        val responseBody = getResponseBodyAsString()

        return if (status == HTTP_OK && responseBody != null) {
            Timber.d("Successful response $responseBody")
            
            // Parse the response
            val moshi: Moshi = Moshi.Builder().build()
            val jsonAdapter: JsonAdapter<TokenResponse> = moshi.adapter(TokenResponse::class.java)
            val tokenResponse: TokenResponse? = jsonAdapter.fromJson(responseBody)
            Timber.d("Get tokens completed and parsed to $tokenResponse")

           RemoteOperationResult<TokenResponse>(RemoteOperationResult.ResultCode.OK).apply {
               data = tokenResponse
           }
       } else {
           Timber.e("Failed response while getting tokens from the server status code: $status; response message: $responseBody")
            RemoteOperationResult(this)
        }
    }
}
```

Reply:
More functional 🍻

Reply:
Yes, I agree that it is more functional, but sometimes I prefer readability. I think it is easier to read this way.

Reply:
Yep, maybe the `with(tokenRequestParams) {` makes it less readable but without that I like how the post method execution looks like

```
PostMethod(
        url = URL(tokenRequestParams tokenEndpoint),
        postRequestBody = tokenRequestParams.toRequestBody()
    ).apply {
        addRequestHeader(AUTHORIZATION_HEADER, clientAuth)
    }.run {
```

But of course, it's a suggestion, I try to improve my code by reducing immutable variables and transforming them to functional, your way is fine as well.